src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,444 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
+    let alloc_arg2 = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "alloc arg2"), store, || {
+        Ok(&arg2_u64_ptr)
+    })?;
+
+    // a = b * q + r
+    let product_u64mod = mul(
+        &mut cs.namespace(|| "product(q,arg2)"),
+        alloc_q.hash(),
+        alloc_arg2.hash(),
+    )?;
+    let sum_u64mod = add(
+        &mut cs.namespace(|| "sum remainder mod u64"),
+        &product_u64mod,
+        alloc_r.hash(),
+    )?;
+    let u64mod_decomp = alloc_equal(
+        &mut cs.namespace(|| "check u64 mod decomposition"),
+        &sum_u64mod,
+        alloc_arg1.hash(),
+    )?;
+    let b_is_zero = alloc_is_zero(&mut cs.namespace(|| "b is zero"), arg2.hash())?;
+    let b_is_not_zero_and_cond = Boolean::and(
+        &mut cs.namespace(|| "b is not zero and cond"),
+        &b_is_zero.not(),
+        &cond,
+    )?;
+    enforce_implication(
+        &mut cs.namespace(|| "enforce u64 mod decomposition"),
+        &b_is_not_zero_and_cond,
+        &u64mod_decomp,
+    )?;
+
+    enforce_less_than_bound(
+        &mut cs.namespace(|| "remainder in range b"),
+        cond,
+        alloc_r.clone(),
+        alloc_arg2.clone(),
+    )?;
+
+    let arg1_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg1 u64 tag"),
+        alloc_arg1.tag(),
+        &g.u64_tag,
+    )?;
+    let arg2_u64_tag = alloc_equal(
+        &mut cs.namespace(|| "arg2 u64 tag"),
+        alloc_arg2.tag(),
+        &g.u64_tag,
+    )?;
+    enforce_true(&mut cs.namespace(|| "check arg1 u64 tag"), &arg1_u64_tag)?;
+    enforce_true(&mut cs.namespace(|| "check arg2 u64 tag"), &arg2_u64_tag)?;
+
+    Ok((alloc_q, alloc_r))
+}
+
+// Enforce the num < b. This is done by proving (b - num) is positive.
+pub fn enforce_less_than_bound<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    cond: Boolean,
+    num: AllocatedPtr<F>,
+    bound: AllocatedPtr<F>,
+) -> Result<(), SynthesisError> {
+    let diff_bound_num = sub(
+        &mut cs.namespace(|| "bound minus num"),
+        bound.hash(),
+        num.hash(),
+    )?;
+    let double_diff_bound_num = add(
+        &mut cs.namespace(|| "double diff bound num"),
+        &diff_bound_num,
+        &diff_bound_num,
+    )?;
+    let double_diff_bound_num_bits = double_diff_bound_num
+        .to_bits_le(&mut cs.namespace(|| "diff bits"))
+        .unwrap();
+    let lsb_2diff_bound_num = double_diff_bound_num_bits.get(0);
+    let diff_bound_num_is_negative = lsb_2diff_bound_num.unwrap();
+
+    enforce_implication(
+        &mut cs.namespace(|| "enforce u64 range"),
+        &cond,
+        &diff_bound_num_is_negative.not(),
+    )
+}
+
+// Convert from bn to num. This allocation is NOT constrained here.
+// In the circuit we use it to prove u64 decomposition, since using bn
+// we have division with remainder, which is used to find the quotient
+// after dividing by 2ˆ64. Therefore we constrain this relation afterwards.
+pub fn bn_to_num<F: LurkField, CS: ConstraintSystem<F>>(
```

Comment:
Let's make very clear what's going on here, including with a flag that this is 'unsafe' (in that it performs an unconstrained allocation). We could name it something like `allocate_unconstrainged_bignum`.

Reply:
done